apps/api/src/app/shared/services/distributed-lock/distributed-lock.service.ts

```diff
@@ -0,0 +1,149 @@
+import Redis from 'ioredis';
+import * as Redlock from 'redlock';
+import { setTimeout } from 'timers/promises';
+import { Logger } from '@nestjs/common';
+
+const LOG_CONTEXT = 'DistributedLock';
+
+const redisUrl =
+  process.env.REDIS_HOST && process.env.REDIS_PORT
+    ? `${process.env.REDIS_HOST}:${process.env.REDIS_PORT}`
+    : 'localhost:6379';
+
+interface ILockOptions {
+  resource: string;
+  ttl: number;
+}
+
+export class DistributedLockService {
+  public distributedLock: Redlock;
+  public instances: Redis[];
+  public lockCounter = {};
```

Comment:
Because the system is distributed and this will be on each instance of the service, what is the user of this lockCounter? Should we maybe remove it?

Reply:
The `lockCounter` is needed for the shutdown operation to avoid the system to shutdown while a lock is in operation (safety measure). What we need to do, and I think it got lost in my previous comments, is to know where I can engage the `shutdown` functionality in our Nest app, because I haven't found any way where we do operations before app Nest shutdown. Can you point me where we could do that?